### PR TITLE
[RFC] the fcn-xs example for image segmentation

### PR DESCRIPTION
**Pull request by tornadomeet**
this is the refactor version of https://github.com/dmlc/mxnet/pull/940, and it  need the change of mshadow, refer the pr https://github.com/dmlc/mshadow/pull/87

### CODE REVIEW

**Comment by tqchen** [comment]
@winstywang can you review this?

**Comment by tqchen** [comment]
One question: Can we reuse(or enhance) the model.fit for fcn training? instead of using the current solver?

**Comment by tornadomeet** [comment]
@tqchen , currently cannot use the model.fit, the biggist problem is: the input size of the images may not be the same, and we should bind at every batch(here the minibatch eques 1, and i only replace the data and label during ervery bind for efficiency), model.fit assume that our input data size is the same during training.
another reason is the mxnet does not support image segmentation io, so no appropriate dataiter passed in model.fit; 
it will be great if mxnet support the image segmentation io (c++ side) for the distributed training in the future!! i just done some work before, but it seems a bit hard for me to add it.
the example here is referenced with https://github.com/BVLC/caffe/wiki/Model-Zoo#fully-convolutional-semantic-segmentation-models-fcn-xs

File: example/fcn-xs/README.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+we have trained a simple fcn-xs model, the parameter is below:
+
+| model | lr (fixed) | epoch |
+| ---- | ----: | ---------: |
+| fcn-32s | 1e-10 | 31 |
+| fcn-16s | 1e-12 | 27 |
+| fcn-8s | 1e-14 | 19 |
+
+the training image number is only : 2027, and the Validation image number is: 462
```
Comment (winstywang) [inline example/fcn-xs/README.md]: Do you have results on VOC submission server? We had better add the number in.

File: example/fcn-xs/data.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    def _read_img(self, img_name, label_name):
+        img = Image.open(os.path.join(self.root_dir, img_name))
+        label = Image.open(os.path.join(self.root_dir, label_name))
+        assert img.size == label.size
+        img = np.array(img, dtype=np.float32)  # (h, w, c)
+        label = np.array(label)  # (h, w)
+        if self.cut_off_size is not None:
+            max_hw = max(img.shape[0], img.shape[1])
+            min_hw = min(img.shape[0], img.shape[1])
+            if max_hw > cut_off_size:
```
Comment (winstywang) [inline example/fcn-xs/data.py]: What if min_hw > cut_off_size?

File: example/fcn-xs/data.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+import numpy as np
+import sys, os
+from mxnet.io import DataIter
+from skimage import io
+from PIL import Image
+
+class FileIter(DataIter):
+    """FileIter object in fcn-xs example. Taking a file list file to get dataiter.
+    in this example, we use the whole image training for fcn-xs, that is to say
+    we do not need resize/crop the image to the same size, so the batch_size is
```
Comment (winstywang) [inline example/fcn-xs/data.py]: For efficiency, I suggest to change to crop image instead, but not required :)

File: example/fcn-xs/data.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    cut_off_size : int
+        if the maximal size of one image is larger than cut_off_size, then it will
+        crop the image with the minimal size of that image
+    data_name : string
+        the data name used in symbol data(default data name)
+    label_name : string
+        the label name used in symbol softmax_label(default label name)
+    """
+    def __init__(self, root_dir, flist_name,
+                 rgb_mean = (117, 117, 117),
```
Comment (winstywang) [inline example/fcn-xs/data.py]: If you use VGG net, the mean is not (117,117,117)

File: example/fcn-xs/fcn_xs.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        )
+    model = Solver(
+        ctx                 = ctx,
+        symbol              = fcnxs,
+        begin_epoch         = 0,
+        num_epoch           = 50,
+        arg_params          = fcnxs_args,
+        aux_params          = fcnxs_auxs,
+        learning_rate       = 1e-10,
+        momentum            = 0.99,
```
Comment (winstywang) [inline example/fcn-xs/fcn_xs.py]: Do we need such a large momentum for training?

File: example/fcn-xs/init_fcnxs.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            v.copyto(fcnxs_args[k])
+    for k,v in fcnxs_auxs.items():
+        if(v.context != ctx):
+            fcnxs_auxs[k] = mx.nd.zeros(v.shape, ctx)
+            v.copyto(fcnxs_auxs[k])
+    data_shape=(1,3,500,500)
+    arg_names = fcnxs_symbol.list_arguments()
+    arg_shapes, _, _ = fcnxs_symbol.infer_shape(data=data_shape)
+    rest_params = dict([(x[0], mx.nd.zeros(x[1], ctx)) for x in zip(arg_names, arg_shapes)
+            if x[0] in ['score_weight', 'score_bias', 'score_pool4_weight', 'score_pool4_bias', \
```
Comment (winstywang) [inline example/fcn-xs/init_fcnxs.py]: Why the weights are initialized to zero not random value?

File: example/fcn-xs/solver.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        self.aux_params = aux_params
+        self.optimizer = optimizer
+        self.kwargs = kwargs.copy()
+
+    def fit(self, train_data, eval_data=None,
+            eval_metric='acc',
+            grad_req='write',
+            epoch_end_callback=None,
+            batch_end_callback=None,
+            kvstore='local',
```
Comment (winstywang) [inline example/fcn-xs/solver.py]: Actually, no kvstore is used?

Comment (winstywang) [inline example/fcn-xs/solver.py]: We can leave it as it is now. I will fix the multi-gpu later.

File: src/operator/crop-inl.h [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+}  // namespace crop_enum
+
+struct CropParam : public dmlc::Parameter<CropParam> {
+  TShape offset;
+  bool center_crop;
+  DMLC_DECLARE_PARAMETER(CropParam) {
+    int shape[] = {0, 0};
+    DMLC_DECLARE_FIELD(offset).set_default(TShape(shape, shape + 2))
+    .describe("corp offset coordinate: (y, x)");
+    DMLC_DECLARE_FIELD(center_crop).set_default(false)
```
Comment (winstywang) [inline src/operator/crop-inl.h]: It is a very strange that we can only use crop_like other than specifying the size explicitly. I suggest to add in these parameters.

File: example/fcn-xs/symbol_fcnxs.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    pool4 = vgg16_pool4(pool3, workspace_default)
+    score = vgg16_score(pool4, numclass, workspace_default)
+    # score 2X
+    score2 = mx.symbol.Deconvolution(data=score, kernel=(4, 4), stride=(2, 2),num_filter=numclass,
+                workspace=workspace_default, name="score2")  # 2X
+    score_pool4 = mx.symbol.Convolution(data=pool4, kernel=(1, 1), num_filter=numclass,
+                workspace=workspace_default, name="score_pool4")
+    score_pool4c = mx.symbol.Crop(data=score_pool4, crop_like=score2,
+                offset=offset()["score_pool4c"], name="score_pool4c")
+    score_fused = mx.symbol.ElementWiseSum(*[score2, score_pool4c], name='score_fused')
```
Comment (winstywang) [inline example/fcn-xs/symbol_fcnxs.py]: You can use + directly here.

File: example/fcn-xs/init_fcnxs.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            fcnxs_auxs[k] = mx.nd.zeros(v.shape, ctx)
+            v.copyto(fcnxs_auxs[k])
+    data_shape=(1,3,500,500)
+    arg_names = fcnxs_symbol.list_arguments()
+    arg_shapes, _, _ = fcnxs_symbol.infer_shape(data=data_shape)
+    rest_params = {}
+    deconv_params = {}
+    # this is fcn8s init from fcn16s
+    if 'score_pool3_weight' in arg_names:
+        rest_params = dict([(x[0], mx.nd.zeros(x[1], ctx)) for x in zip(arg_names, arg_shapes)
```
Comment (winstywang) [inline example/fcn-xs/init_fcnxs.py]: Again, not zero here.

File: example/fcn-xs/init_fcnxs.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+            if x[0] in ['score_pool3_bias', 'score_pool3_weight']])
+        deconv_params = dict([(x[0], x[1]) for x in zip(arg_names, arg_shapes) if x[0] \
+            in ["bigscore_weight", 'score4_weight']])
+    # this is fcn16s init from fcn32s
+    elif 'score_pool4_weight' in arg_names:
+        rest_params = dict([(x[0], mx.nd.zeros(x[1], ctx)) for x in zip(arg_names, arg_shapes)
```
Comment (winstywang) [inline example/fcn-xs/init_fcnxs.py]: Again, not zero here.

**Comment by winstywang** [comment]
I have reviewed the codes. The codes are generally good. One major issue is that it uses different size for training. Actually, we can randomly crop the input to a fixed size instead. It will speed up the training significantly, and make the codes shorter and cleaner.

File: example/fcn-xs/README.md (hunk 1 above)
Comment (tornadomeet) [inline example/fcn-xs/README.md]: It has not yet.
in this example, i used the sub-set of pascal voc dataset, and random choose 70% number images for training, it seems that it is not the standard train,val.
so if evaluate on voc submission server, i need to retrain the model, and after the new year hodlday, i'll do it, so please wait me for about 2 weeks(one week for train:) )

File: example/fcn-xs/fcn_xs.py (hunk 5 above)
Comment (tornadomeet) [inline example/fcn-xs/fcn_xs.py]: i followed the caffe momentum for setting, and the autor said it will be good to use larger in image segmentation, and i do not change it for training, so maybe we can adjust it.

File: example/fcn-xs/init_fcnxs.py (hunk 6 above)
Comment (tornadomeet) [inline example/fcn-xs/init_fcnxs.py]: yes, it a trick here~  because the class label 0 is the  background in image segmentation(general setting), and it has many background in one image, so if we init those weight to zeros, then at the first time it tend to ouput 0, and the back error will be large, and it benefit for the training.
here is the discuss about the initialize, https://groups.google.com/forum/#!msg/caffe-users/bsGdmJQ67Ts/tyy0zq2Cl28J , there is some answer from the  author which implentating the caffe version of fcn-xs

File: src/operator/crop-inl.h (hunk 8 above)
Comment (tornadomeet) [inline src/operator/crop-inl.h]: yes, it is a bit hard for explicitly the size, the reason is that we cannot konw the crop size before we do the crop operator, because it depends on the input image size(fcn-32s) and the intermdiate layer size(fcn-16s and fcn-8s), and in caffe, https://github.com/BVLC/caffe/pull/1976 it add more code for the crop layer.
the crop layer is used for algiment between two layer, like in spp-net, the proposal box between the input layer and conv5 feature map, and in spp-net, it used a special crop like layer doing the algiment.
the crop layer in caffe is also need the auto alignment in https://github.com/longjon/caffe/blob/future/src/caffe/layers/crop_layer.cpp#L13

ps: in the future, we should add the value of crop in the symbol side, instead of  hand calculation in this example.

File: example/fcn-xs/init_fcnxs.py (hunk 11 above)
Comment (tornadomeet) [inline example/fcn-xs/init_fcnxs.py]: for the above explain~

File: example/fcn-xs/solver.py (hunk 7 above)
Comment (tornadomeet) [inline example/fcn-xs/solver.py]: yes, the batch-size is 1 here, no kvstore.
the minibatch version training will add in the further~

File: example/fcn-xs/data.py (hunk 2 above)
Comment (tornadomeet) [inline example/fcn-xs/data.py]: it's a bug, i'll fix it tomorrow, ths~

**Comment by tornadomeet** [comment]
@winstywang , thanks, i have answer some corresponding review, and i'll fix the cut_off_size bug first.

File: example/fcn-xs/symbol_fcnxs.py (hunk 9 above)
Comment (tornadomeet) [inline example/fcn-xs/symbol_fcnxs.py]: cool, i'll try it.

File: example/fcn-xs/README.md (hunk 1 above)
Comment (winstywang) [inline example/fcn-xs/README.md]: Cool, after merge in this PR, I could do this.

File: src/operator/crop-inl.h (hunk 8 above)
Comment (winstywang) [inline src/operator/crop-inl.h]: No, I mean you could add in the parameters, then we can reuse the op in other applications. Moreover, if we crop the input to the same size, it seems that we don't need it...

File: example/fcn-xs/init_fcnxs.py (hunk 6 above)
Comment (winstywang) [inline example/fcn-xs/init_fcnxs.py]: OK

File: example/fcn-xs/solver.py (hunk 7 above)
Comment (winstywang) [inline example/fcn-xs/solver.py]: You could refer to the DeepLab's codes to implement this.

**Comment by tornadomeet** [comment]
@winstywang , i update the example by:
1. fix the cut_off_size bug of data.py
2. change ElementWiseSum to directly "+" symbol.
3. add the h_w parameter for Crop Operator, and this can be used for general cropping.
4. fix/add some description.
5. update the new symbol in {yun.baidu](http://pan.baidu.com/s/1jGlOvno) for used.

ps, thought i add the h_w for explicitly set the height and width for crop, but i also retain the use of "crop_like" symbol for height and width when corpping. the reason is that even if we crop the input data with the same size, we should also need to manually calculator the crop h_w in fcn16s and fcn8s, just like in https://github.com/tornadomeet/mxnet/blob/b8b9700fcdb1d7fb4c79a3d271da178639453b8a/example/fcn-xs/symbol_fcnxs.py#L179 and https://github.com/tornadomeet/mxnet/blob/b8b9700fcdb1d7fb4c79a3d271da178639453b8a/example/fcn-xs/symbol_fcnxs.py#L186 , we should now the size of score_pool4c and score_pool3c, so if some hyper-parameter of cnn structure is changed, like the stride, pad... then we should manually calculator the size again, so the crop_like symbol is used for auto calculate the crop_size without consider the change of paramter.
what do you think about this?

File: src/operator/crop-inl.h [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+                  std::vector<TShape> *aux_shape) const override {
+    using namespace mshadow;
+    CHECK_EQ(in_shape->size(), static_cast<size_t>(param_.num_args));
+    TShape data_shape = in_shape->at(crop_enum::kData);
+    if (data_shape.ndim() == 0) return false;
+    CHECK_EQ(data_shape.ndim(), 4) << \
+        "Input data should be 4D in batch-num_filter-y-x";
+    std::vector<int> crop_shape;
+    if (param_.num_args == 1) {
+      std::cout << "ok1" << std::endl;
```
Comment (winstywang) [inline src/operator/crop-inl.h]: Remove this debug info

File: example/fcn-xs/data.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+            if min_hw > self.cut_off_size:
+                rand_start_max = round(np.random.uniform(0, max_hw - self.cut_off_size - 1))
+                rand_start_min = round(np.random.uniform(0, min_hw - self.cut_off_size - 1))
+                if img.shape[0] == max_hw :
+                    img = img[rand_start_max : rand_start_max + self.cut_off_size, rand_start_min : rand_start_min + self.cut_off_size]
+                    label = label[rand_start_max : rand_start_max + self.cut_off_size, rand_start_min : rand_start_min + self.cut_off_size]
+                else :
+                    img = img[rand_start_min : rand_start_min + self.cut_off_size, rand_start_max : rand_start_max + self.cut_off_size]
+                    label = label[rand_start_min : rand_start_min + self.cut_off_size, rand_start_max : rand_start_max + self.cut_off_size]
+            elif max_hw > self.cut_off_size:
```
Comment (winstywang) [inline example/fcn-xs/data.py]: The logic is still wired here. What if both min_hw and max_hw is larger than cut_off_size?

**Comment by winstywang** [comment]
Two more comments @tornadomeet 
In this application, using crop_like is proper. What I mean is that in implementation of new op, we had better make it more general. Current implementation is good.

After fixing these two issues, we can merge in this PR, though some further efforts are needed to make it better, including but not limited to:

1) Using batch for training.
2) Multi-GPU support
3) Add evaluation on VOC

**Comment by tornadomeet** [comment]
@winstywang , ths~ this example should be enhance more in in further, especially the batch training.

File: src/operator/softmax_output-inl.h [hunk 14]
```diff
@@ -88,8 +96,12 @@ class SoftmaxOutputOp : public Operator {
...
       Tensor<xpu, 3> out = out_data[softmaxout_enum::kOut].get_with_shape<xpu, 3, real_t>(s3, s);
       Tensor<xpu, 3> grad = in_grad[softmaxout_enum::kData].get_with_shape<xpu, 3, real_t>(s3, s);
-      SoftmaxGrad(grad, out, label);
-      grad *= param_.grad_scale/s3[2];
+      if (param_.use_ignore) {
+          SoftmaxGrad(grad, out, label, static_cast<real_t>(param_.ignore_label));
+      } else {
+          SoftmaxGrad(grad, out, label);
+      }
+      grad *= param_.grad_scale;
```
Comment (piiswrong) [inline src/operator/softmax_output-inl.h]: @tornadomeet Why did you remove the scaling by 1/s3[2]? This causes gradient to explode for large output map.

Comment (tornadomeet) [inline src/operator/softmax_output-inl.h]: @piiswrong , sorry for that, i remove it by accidently, could you please recover it, i'm not in my computer today, ths~

**Comment by zeakey** [comment]
@tornadomeet @winstywang 
I run the fcn demo but error raised:
`raise ValueError('Must specify all the arguments in %s' % arg_key)`
in `python/mexnet/sombol.py` line585, cause name "bigscore_bias" is not in "arg_names".

I wonder if this is a bug or just I haven't set `fcnxs_model_prefix` properly ?

**Comment by tornadomeet** [comment]
@zeakey https://github.com/dmlc/mxnet/pull/1627

**Comment by zeakey** [comment]
@tornadomeet OK, thank you for all !  It's boring for people new to mxnet
